devito/propagator.py

```diff
@@ -30,10 +33,20 @@ def __init__(self, name, nt, shape, spc_border=0, forward=True, time_order=0):
         # Start with the assumption that the propagator needs to save the field in memory at every time step
         self._save = True
         # This might be changed later when parameters are being set
-
+        self.profile = os.environ.get(self._ENV_VAR_PROFILE) == "1"
         # Which function parameters need special (non-save) time stepping and which don't
         self.save_vars = {}
         self.fd = FunctionDescriptor(name)
+        self.pre_loop = []
+        self.post_loop = []
+        if self.profile:
+            self.add_local_var("time", "double")
+            self.pre_loop.append(cgen.Statement("struct timeval start, end"))
+            self.pre_loop.append(cgen.Assign("time", 0))
+            self.post_loop.append(cgen.PrintStatement("time: %f\n", "time"))
+            self.time_loop_stencils_b.append(cgen.Statement("gettimeofday(&start, NULL)"))
+            self.time_loop_stencils_a.append(cgen.Statement("gettimeofday(&end, NULL)"))
+            self.time_loop_stencils_a.append(cgen.Statement("time += ((end.tv_sec  - start.tv_sec) * 1000000u + end.tv_usec - start.tv_usec) / 1.e6"))
```

Comment:
The problem is that if we move it out we are then profiling the addition of the source term. This is not a hot spot but adding it's time at this point blurs the picture if performance for the main loop.

Then again - perhaps as it is small it then it could be only a percentage point... How about someone test this for the 256**3 case and move it out if the difference is small?